File: src/retrocast/v2/adapters/askcos.py
from __future__ import annotations

from collections import defaultdict
from collections.abc import Iterator, Mapping
from dataclasses import dataclass
from types import MappingProxyType
from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field, ValidationError

from retrocast.adapters.errors import (
    adapter_cycle_error,
    adapter_missing_node_error,
    adapter_schema_error,
    adapter_target_mismatch,
)
from retrocast.chem import canonicalize_smiles, get_inchi_key
from retrocast.exceptions import AdapterLogicError, InvalidSmilesError, UnsupportedAdapterFeatureError
from retrocast.typing import ReactionSmilesStr, SmilesStr
from retrocast.v2.adapters.base import AdaptMode, RawRouteEntry
from retrocast.v2.models.route import Molecule, Reaction, Route
from retrocast.v2.models.task import Target

ASKCOS_ROOT_UUID = "00000000-0000-0000-0000-000000000000"


# SECTION: Raw ASKCOS Schema


class AskcosBaseNode(BaseModel):
    smiles: str
    id: str


class AskcosChemicalNode(AskcosBaseNode):
    type: Literal["chemical"]
    terminal: bool


class AskcosModelMetadata(BaseModel):
    source: dict[str, Any] = Field(default_factory=dict)

    def get_template(self) -> str | None:
        template = self.source.get("template")
        if not isinstance(template, dict):
            return None
        reaction_smarts = template.get("reaction_smarts")
        return reaction_smarts if isinstance(reaction_smarts, str) else None


class AskcosReactionProperties(BaseModel):
    mapped_smiles: str | None = None


class AskcosReactionNode(AskcosBaseNode):
    type: Literal["reaction"]
    reaction_properties: AskcosReactionProperties | None = None
    model_metadata: list[AskcosModelMetadata] = Field(default_factory=list)


AskcosNode = Annotated[AskcosChemicalNode | AskcosReactionNode, Field(discriminator="type")]


class AskcosPathwayEdge(BaseModel):
    source: str
    target: str


class AskcosUDS(BaseModel):
    node_dict: dict[str, AskcosNode]
    uuid2smiles: dict[str, str]
    pathways: list[list[AskcosPathwayEdge]]


class AskcosResults(BaseModel):
    uds: AskcosUDS

    stats: dict[str, Any] = Field(default_factory=dict)


class AskcosOutput(BaseModel):
    results: AskcosResults


@dataclass(slots=True)
class AskcosPathwayPayload:
    pathway_edges: tuple[AskcosPathwayEdge, ...]
    uuid2smiles: MappingProxyType[str, str]
    node_dict: MappingProxyType[str, AskcosNode]
    annotations: MappingProxyType[str, Any]


# SECTION: Adapter


class AskcosAdapter:
    def __init__(self, use_full_graph: bool = False) -> None:
        self.use_full_graph = use_full_graph

    def iter_raw_routes(
        self,
        raw_payload: Any,
        *,
        source_key: str | None = None,
    ) -> Iterator[RawRouteEntry]:
        if self.use_full_graph:
            raise UnsupportedAdapterFeatureError(
                "ASKCOS full-graph route extraction is not implemented",
                context={"adapter": "askcos", "feature": "full_graph"},
            )

        target_id = source_key or "<unknown>"
        try:
            output = AskcosOutput.model_validate(raw_payload)
        except ValidationError as exc:
            raise adapter_schema_error("askcos", target_id, "invalid output") from exc

        annotations = _extract_run_annotations(output.results.stats)
        uds = output.results.uds
        for pathway_index, pathway_edges in enumerate(uds.pathways, start=1):
            yield RawRouteEntry(
                payload=AskcosPathwayPayload(
                    pathway_edges=tuple(pathway_edges),
                    uuid2smiles=MappingProxyType(uds.uuid2smiles),
                    node_dict=MappingProxyType(uds.node_dict),
                    annotations=MappingProxyType(annotations),
                ),
                source_key=source_key,
                source_order=pathway_index,
            )

    def cast(
        self,
        raw_route: Any,
        *,
        mode: AdaptMode = "strict",
        target: Target | None = None,
    ) -> Route:
        target_id = target.id if target is not None else "<unknown>"
        if not isinstance(raw_route, AskcosPathwayPayload):
            raise adapter_schema_error("askcos", target_id, "invalid pathway")

        route_target = self._build_target_molecule(raw_route, mode=mode)
        if route_target is None:
            raise AdapterLogicError(
                "ASKCOS target molecule was pruned",
                code="adapter.target_pruned",
                context={"adapter": "askcos", "target_id": target_id},
            )

        if target is not None:
            expected_smiles = canonicalize_smiles(target.smiles)
            if route_target.smiles != expected_smiles:
                raise adapter_target_mismatch(
                    "askcos",
                    target.id,
                    expected_smiles=expected_smiles,
                    actual_smiles=route_target.smiles,
                )

        return Route(target=route_target, annotations=raw_route.annotations)

    def _build_target_molecule(self, raw_route: AskcosPathwayPayload, *, mode: AdaptMode) -> Molecule | None:
        adj_list: dict[str, list[str]] = defaultdict(list)
        for edge in raw_route.pathway_edges:
            adj_list[edge.source].append(edge.target)

        if ASKCOS_ROOT_UUID not in raw_route.uuid2smiles:
            raise adapter_missing_node_error(
                "askcos", node_id=ASKCOS_ROOT_UUID, lookup="uuid2smiles", role="root chemical"
            )

        return self._build_molecule(
            chem_uuid=ASKCOS_ROOT_UUID,
            adj_list=adj_list,
            uuid2smiles=raw_route.uuid2smiles,
            node_dict=raw_route.node_dict,
            visited=set(),
            mode=mode,
        )

    def _build_molecule(
        self,
        *,
        chem_uuid: str,
        adj_list: dict[str, list[str]],
        uuid2smiles: Mapping[str, str],
        node_dict: Mapping[str, AskcosNode],
        visited: set[SmilesStr],
        mode: AdaptMode,
    ) -> Molecule | None:
        raw_smiles = uuid2smiles.get(chem_uuid)
        if not raw_smiles:
            raise adapter_missing_node_error("askcos", node_id=chem_uuid, lookup="uuid2smiles", role="chemical")

        node = node_dict.get(raw_smiles)
        if not isinstance(node, AskcosChemicalNode):
            raise adapter_missing_node_error("askcos", node_id=raw_smiles, lookup="node_dict", role="chemical")

        try:
            canon_smiles = canonicalize_smiles(node.smiles)
        except InvalidSmilesError:
            if mode == "prune":
                return None
            raise

        if canon_smiles in visited:
            raise adapter_cycle_error("askcos", canon_smiles)

        if node.terminal or chem_uuid not in adj_list:
            return Molecule(smiles=canon_smiles, inchikey=get_inchi_key(canon_smiles))

        child_reaction_uuids = adj_list[chem_uuid]
        if len(child_reaction_uuids) > 1:
            raise AdapterLogicError(
                "ASKCOS pathway is not a route tree: molecule has multiple child reactions",
                code="adapter.route_not_tree",
                context={
                    "adapter": "askcos",
                    "smiles": canon_smiles,
                    "child_reaction_count": len(child_reaction_uuids),
                },
            )

        reaction = self._build_reaction(
            rxn_uuid=child_reaction_uuids[0],
            adj_list=adj_list,
            uuid2smiles=uuid2smiles,
            node_dict=node_dict,
            visited=visited | {canon_smiles},
            mode=mode,
        )
        if reaction is None:
            if mode == "prune":
                return None
            raise AdapterLogicError(
                "ASKCOS reaction has no reactants",
                code="adapter.reaction_empty",
                context={"adapter": "askcos", "smiles": canon_smiles},
            )

        return Molecule(smiles=canon_smiles, inchikey=get_inchi_key(canon_smiles), product_of=reaction)

    def _build_reaction(
        self,
        *,
        rxn_uuid: str,
        adj_list: dict[str, list[str]],
        uuid2smiles: Mapping[str, str],
        node_dict: Mapping[str, AskcosNode],
        visited: set[SmilesStr],
        mode: AdaptMode,
    ) -> Reaction | None:
        raw_smiles = uuid2smiles.get(rxn_uuid)
        if not raw_smiles:
            raise adapter_missing_node_error("askcos", node_id=rxn_uuid, lookup="uuid2smiles", role="reaction")

        node = node_dict.get(raw_smiles)
        if not isinstance(node, AskcosReactionNode):
            raise adapter_missing_node_error("askcos", node_id=raw_smiles, lookup="node_dict", role="reaction")

        reactants: list[Molecule] = []
        for reactant_uuid in adj_list.get(rxn_uuid, []):
            reactant = self._build_molecule(
                chem_uuid=reactant_uuid,
                adj_list=adj_list,
                uuid2smiles=uuid2smiles,
                node_dict=node_dict,
                visited=visited,
                mode=mode,
            )
            if reactant is not None:
                reactants.append(reactant)

        if not reactants:
            return None

        mapped_reaction_smiles = None
        if node.reaction_properties and node.reaction_properties.mapped_smiles:
            mapped_reaction_smiles = ReactionSmilesStr(node.reaction_properties.mapped_smiles)

        return Reaction(
            reactants=reactants,
            mapped_reaction_smiles=mapped_reaction_smiles,
            template=node.model_metadata[0].get_template() if node.model_metadata else None,
            annotations={"source_id": node.id},
        )


# SECTION: Annotations


def _extract_run_annotations(stats: dict[str, Any]) -> dict[str, Any]:
    return {
        key: stats[key]
        for key in ("total_iterations", "total_chemicals", "total_reactions", "total_templates", "total_paths")
        if key in stats
    }

File: packages/retrocast-rs/crates/retrocast-core/src/adapters/askcos.rs
use std::collections::{BTreeMap, HashSet};

use rayon::prelude::*;
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value, json};

use super::{Adapter, RawRouteEntry, candidate_from_result, common};
use crate::{
    chem,
    error::Result,
    model::{Candidate, Molecule, Reaction, Route, Target},
    route::{AdaptMode, normalize_reactants},
    schema::{CanonicalSmiles, ReactionSmiles},
    with_pool,
};

const ROOT_UUID: &str = "00000000-0000-0000-0000-000000000000";

#[derive(Clone, Debug, Deserialize, Serialize)]
struct Edge {
    source: String,
    target: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
struct ExtractedPathway {
    pathway_edges: Vec<Edge>,
    uuid2smiles: BTreeMap<String, String>,
    node_dict: BTreeMap<String, Value>,
    annotations: Map<String, Value>,
}

struct AskcosOutput {
    pathways: Vec<Vec<Edge>>,
    uuid2smiles: BTreeMap<String, String>,
    node_dict: BTreeMap<String, Value>,
    annotations: Map<String, Value>,
}

#[derive(Clone, Copy, Debug, Default)]
pub struct AskcosAdapter {
    use_full_graph: bool,
}

impl AskcosAdapter {
    pub fn full_graph() -> Self {
        Self {
            use_full_graph: true,
        }
    }
}

impl Adapter for AskcosAdapter {
    fn name(&self) -> &'static str {
        "askcos"
    }

    fn entries(&self, payload: Value, source_key: Option<&str>) -> Result<Vec<RawRouteEntry>> {
        let output = parse_output(payload, self)?;
        Ok(output
            .pathways
            .into_iter()
            .enumerate()
            .map(|(index, edges)| RawRouteEntry {
                payload: serde_json::to_value(compact_pathway(
                    edges,
                    &output.uuid2smiles,
                    &output.node_dict,
                    &output.annotations,
                ))
                .expect("ASKCOS pathway is serializable"),
                source_key: source_key.map(str::to_owned),
                source_row_index: None,
                source_record_id: None,
                target_hint_id: None,
                target_hint_smiles: None,
                source_order: Some(index + 1),
            })
            .collect())
    }

    #[allow(clippy::too_many_arguments)]
    fn candidates(
        &self,
        payload: Value,
        mode: AdaptMode,
        target: Option<&Target>,
        _source_key: Option<&str>,
        max_candidates: Option<usize>,
        workers: usize,
    ) -> Result<Vec<Candidate>> {
        let AskcosOutput {
            mut pathways,
            uuid2smiles,
            node_dict,
            annotations,
        } = parse_output(payload, self)?;
        if let Some(limit) = max_candidates {
            pathways.truncate(limit);
        }

        let adapt = |(index, edges): (usize, Vec<Edge>)| {
            let rank = index + 1;
            let entry = RawRouteEntry::new(Value::Null);
            candidate_from_result(
                self.name(),
                rank,
                build_route(&edges, &uuid2smiles, &node_dict, &annotations, mode, target),
                target,
                &entry,
            )
        };

        if workers == 1 {
            return Ok(pathways.into_iter().enumerate().map(adapt).collect());
        }
        with_pool(workers, || {
            pathways.into_par_iter().enumerate().map(adapt).collect()
        })
    }

    fn cast(&self, raw_route: Value, mode: AdaptMode, target: Option<&Target>) -> Result<Route> {
        let pathway: ExtractedPathway = serde_json::from_value(raw_route)
            .map_err(|error| common::schema(self.name(), format!("invalid pathway: {error}")))?;
        build_route(
            &pathway.pathway_edges,
            &pathway.uuid2smiles,
            &pathway.node_dict,
            &pathway.annotations,
            mode,
            target,
        )
    }
}

fn parse_output(payload: Value, adapter: &AskcosAdapter) -> Result<AskcosOutput> {
    if adapter.use_full_graph {
        return Err(common::logic(
            adapter.name(),
            "adapter.unsupported_feature",
            "full-graph route extraction is not implemented",
        ));
    }
    let Value::Object(mut payload) = payload else {
        return Err(common::schema(adapter.name(), "results object is required"));
    };
    let results = payload
        .remove("results")
        .ok_or_else(|| common::schema(adapter.name(), "results object is required"))?;
    let Value::Object(mut results) = results else {
        return Err(common::schema(adapter.name(), "results object is required"));
    };
    let annotations = run_annotations(results.get("stats"));
    let uds = results
        .remove("uds")
        .ok_or_else(|| common::schema(adapter.name(), "results.uds object is required"))?;
    let Value::Object(mut uds) = uds else {
        return Err(common::schema(
            adapter.name(),
            "results.uds object is required",
        ));
    };
    let uuid2smiles = serde_json::from_value(
        uds.remove("uuid2smiles")
            .ok_or_else(|| common::schema(adapter.name(), "uds.uuid2smiles is required"))?,
    )
    .map_err(|error| common::schema(adapter.name(), format!("invalid uuid2smiles: {error}")))?;
    let node_dict = serde_json::from_value(
        uds.remove("node_dict")
            .ok_or_else(|| common::schema(adapter.name(), "uds.node_dict is required"))?,
    )
    .map_err(|error| common::schema(adapter.name(), format!("invalid node_dict: {error}")))?;
    let pathways = serde_json::from_value(
        uds.remove("pathways")
            .ok_or_else(|| common::schema(adapter.name(), "uds.pathways is required"))?,
    )
    .map_err(|error| common::schema(adapter.name(), format!("invalid pathways: {error}")))?;
    validate_nodes(&node_dict, adapter.name())?;
    Ok(AskcosOutput {
        pathways,
        uuid2smiles,
        node_dict,
        annotations,
    })
}

fn build_route(
    pathway_edges: &[Edge],
    uuid2smiles: &BTreeMap<String, String>,
    node_dict: &BTreeMap<String, Value>,
    annotations: &Map<String, Value>,
    mode: AdaptMode,
    target: Option<&Target>,
) -> Result<Route> {
    let mut adjacency: BTreeMap<&str, Vec<&str>> = BTreeMap::new();
    for edge in pathway_edges {
        adjacency
            .entry(edge.source.as_str())
            .or_default()
            .push(edge.target.as_str());
    }
    if !uuid2smiles.contains_key(ROOT_UUID) {
        return Err(node_missing("root chemical", ROOT_UUID, "uuid2smiles"));
    }
    let target_molecule = build_molecule(
        ROOT_UUID,
        &adjacency,
        uuid2smiles,
        node_dict,
        mode,
        &mut HashSet::new(),
    )?
    .ok_or_else(|| common::target_pruned("askcos"))?;
    if let Some(target) = target {
        common::require_target_match(
            &target_molecule.smiles,
            &target.smiles,
            &target.id,
            "askcos",
        )?;
    }
    Ok(Route {
        target: target_molecule,
        annotations: annotations.clone(),
        schema_version: Default::default(),
    })
}

fn compact_pathway(
    pathway_edges: Vec<Edge>,
    uuid2smiles: &BTreeMap<String, String>,
    node_dict: &BTreeMap<String, Value>,
    annotations: &Map<String, Value>,
) -> ExtractedPathway {
    let pathway_uuids: HashSet<&str> = pathway_edges
        .iter()
        .flat_map(|edge| [edge.source.as_str(), edge.target.as_str()])
        .chain([ROOT_UUID])
        .collect();
    let compact_uuid2smiles: BTreeMap<_, _> = pathway_uuids
        .into_iter()
        .filter_map(|uuid| {
            uuid2smiles
                .get(uuid)
                .map(|smiles| (uuid.to_owned(), smiles.clone()))
        })
        .collect();
    let compact_node_dict = compact_uuid2smiles
        .values()
        .filter_map(|smiles| {
            node_dict
                .get(smiles)
                .map(|node| (smiles.clone(), node.clone()))
        })
        .collect();
    ExtractedPathway {
        pathway_edges,
        uuid2smiles: compact_uuid2smiles,
        node_dict: compact_node_dict,
        annotations: annotations.clone(),
    }
}

fn validate_nodes(nodes: &BTreeMap<String, Value>, adapter: &'static str) -> Result<()> {
    for (key, node) in nodes {
        let object = node.as_object().ok_or_else(|| {
            common::schema(adapter, format!("node_dict[{key:?}] must be an object"))
        })?;
        let kind = object.get("type").and_then(Value::as_str).ok_or_else(|| {
            common::schema(adapter, format!("node_dict[{key:?}].type is required"))
        })?;
        if !matches!(kind, "chemical" | "reaction") {
            return Err(common::schema(
                adapter,
                format!("node_dict[{key:?}] has invalid type {kind:?}"),
            ));
        }
        for field in ["smiles", "id"] {
            if object.get(field).and_then(Value::as_str).is_none() {
                return Err(common::schema(
                    adapter,
                    format!("node_dict[{key:?}].{field} must be a string"),
                ));
            }
        }
        if kind == "chemical" && object.get("terminal").and_then(Value::as_bool).is_none() {
            return Err(common::schema(
                adapter,
                format!("node_dict[{key:?}].terminal must be a bool"),
            ));
        }
    }
    Ok(())
}

fn build_molecule(
    chemical_uuid: &str,
    adjacency: &BTreeMap<&str, Vec<&str>>,
    uuid2smiles: &BTreeMap<String, String>,
    node_dict: &BTreeMap<String, Value>,
    mode: AdaptMode,
    visited: &mut HashSet<CanonicalSmiles>,
) -> Result<Option<Molecule>> {
    let raw_smiles = uuid2smiles
        .get(chemical_uuid)
        .filter(|smiles| !smiles.is_empty())
        .ok_or_else(|| node_missing("chemical", chemical_uuid, "uuid2smiles"))?;
    let node = node_dict
        .get(raw_smiles)
        .and_then(Value::as_object)
        .filter(|node| node.get("type").and_then(Value::as_str) == Some("chemical"))
        .ok_or_else(|| node_missing("chemical", raw_smiles, "node_dict"))?;
    let (smiles, inchikey) = match chem::normalize(
        node.get("smiles")
            .and_then(Value::as_str)
            .expect("validated chemical smiles"),
    ) {
        Ok(identity) => identity,
        Err(_) if mode == AdaptMode::Prune => return Ok(None),
        Err(error) => return Err(error),
    };
    if visited.contains(&smiles) {
        return Err(common::logic(
            "askcos",
            "adapter.cycle_detected",
            format!("cycle at {smiles:?}"),
        ));
    }
    let child_reactions = adjacency
        .get(chemical_uuid)
        .map(Vec::as_slice)
        .unwrap_or_default();
    if node["terminal"].as_bool().expect("validated terminal") || child_reactions.is_empty() {
        return Ok(Some(Molecule {
            smiles,
            inchikey,
            product_of: None,
            annotations: Map::new(),
        }));
    }
    if child_reactions.len() > 1 {
        return Err(common::logic(
            "askcos",
            "adapter.route_not_tree",
            format!("molecule {smiles:?} has multiple child reactions"),
        ));
    }
    visited.insert(smiles.clone());
    let reaction = build_reaction(
        child_reactions[0],
        adjacency,
        uuid2smiles,
        node_dict,
        mode,
        visited,
    );
    visited.remove(&smiles);
    let reaction = reaction?;
    let Some(reaction) = reaction else {
        return if mode == AdaptMode::Prune {
            Ok(None)
        } else {
            Err(common::logic(
                "askcos",
                "adapter.reaction_empty",
                format!("reaction for {smiles:?} has no reactants"),
            ))
        };
    };
    Ok(Some(Molecule {
        smiles,
        inchikey,
        product_of: Some(Box::new(reaction)),
        annotations: Map::new(),
    }))
}

fn build_reaction(
    reaction_uuid: &str,
    adjacency: &BTreeMap<&str, Vec<&str>>,
    uuid2smiles: &BTreeMap<String, String>,
    node_dict: &BTreeMap<String, Value>,
    mode: AdaptMode,
    visited: &mut HashSet<CanonicalSmiles>,
) -> Result<Option<Reaction>> {
    let raw_smiles = uuid2smiles
        .get(reaction_uuid)
        .filter(|smiles| !smiles.is_empty())
        .ok_or_else(|| node_missing("reaction", reaction_uuid, "uuid2smiles"))?;
    let node = node_dict
        .get(raw_smiles)
        .and_then(Value::as_object)
        .filter(|node| node.get("type").and_then(Value::as_str) == Some("reaction"))
        .ok_or_else(|| node_missing("reaction", raw_smiles, "node_dict"))?;
    let mut reactants = Vec::new();
    for reactant_uuid in adjacency
        .get(reaction_uuid)
        .map(Vec::as_slice)
        .unwrap_or_default()
    {
        if let Some(reactant) = build_molecule(
            reactant_uuid,
            adjacency,
            uuid2smiles,
            node_dict,
            mode,
            visited,
        )? {
            reactants.push(reactant);
        }
    }
    if reactants.is_empty() {
        return Ok(None);
    }
    normalize_reactants(&mut reactants);
    let mapped_reaction_smiles = node
        .get("reaction_properties")
        .and_then(Value::as_object)
        .and_then(|properties| properties.get("mapped_smiles"))
        .and_then(Value::as_str)
        .map(ReactionSmiles::try_from)
        .transpose()?;
    Ok(Some(Reaction {
        reactants,
        mapped_reaction_smiles,
        template: template(node),
        reagents: None,
        solvents: None,
        annotations: Map::from_iter([(
            "source_id".to_owned(),
            json!(node["id"].as_str().expect("validated reaction id")),
        )]),
    }))
}

fn template(node: &Map<String, Value>) -> Option<String> {
    node.get("model_metadata")?
        .as_array()?
        .first()?
        .get("source")?
        .get("template")?
        .get("reaction_smarts")?
        .as_str()
        .map(str::to_owned)
}

fn node_missing(role: &str, node_id: &str, lookup: &str) -> crate::error::EngineError {
    common::logic(
        "askcos",
        "adapter.node_missing",
        format!("missing {role} {node_id:?} in {lookup}"),
    )
}

fn run_annotations(stats: Option<&Value>) -> Map<String, Value> {
    let Some(stats) = stats.and_then(Value::as_object) else {
        return Map::new();
    };
    [
        "total_iterations",
        "total_chemicals",
        "total_reactions",
        "total_templates",
        "total_paths",
    ]
    .into_iter()
    .filter_map(|key| stats.get(key).map(|value| (key.to_owned(), value.clone())))
    .collect()
}

#[cfg(test)]
mod tests {
    use serde_json::{Value, json};

    use super::{AskcosAdapter, ExtractedPathway};
    use crate::{
        adapters::{Adapter, adapt_candidates_with_workers},
        route::AdaptMode,
    };

    fn output() -> Value {
        json!({"results": {
            "stats": {"total_paths": 1, "ignored": 2},
            "uds": {
                "node_dict": {
                    "CCOC(C)=O": {"smiles": "CCOC(C)=O", "id": "chem-root", "type": "chemical", "terminal": false},
                    "CCO": {"smiles": "CCO", "id": "chem-leaf", "type": "chemical", "terminal": true},
                    "CO": {"smiles": "CO", "id": "chem-unused", "type": "chemical", "terminal": true},
                    "CCO>>CCOC(C)=O": {
                        "smiles": "CCO>>CCOC(C)=O", "id": "rxn-1", "type": "reaction",
                        "reaction_properties": {"mapped_smiles": "CCO>>CCOC(C)=O"},
                        "model_metadata": [{"source": {"template": {"reaction_smarts": "esterification"}}}]
                    }
                },
                "uuid2smiles": {
                    "00000000-0000-0000-0000-000000000000": "CCOC(C)=O",
                    "uuid-rxn": "CCO>>CCOC(C)=O",
                    "uuid-leaf": "CCO",
                    "uuid-unused": "CO"
                },
                "pathways": [[
                    {"source": "00000000-0000-0000-0000-000000000000", "target": "uuid-rxn"},
                    {"source": "uuid-rxn", "target": "uuid-leaf"}
                ]]
            }
        }})
    }

    #[test]
    fn extracts_path_and_preserves_reaction_fields() {
        let entry = AskcosAdapter::default()
            .entries(output(), None)
            .unwrap()
            .pop()
            .unwrap();
        let pathway: ExtractedPathway = serde_json::from_value(entry.payload).unwrap();
        assert_eq!(pathway.node_dict.len(), 3);
        assert!(!pathway.node_dict.contains_key("CO"));
        let route = AskcosAdapter::default()
            .cast(
                serde_json::to_value(pathway).unwrap(),
                AdaptMode::Strict,
                None,
            )
            .unwrap();
        assert_eq!(route.annotations["total_paths"], 1);
        let reaction = route.target.product_of.unwrap();
        assert_eq!(reaction.template.as_deref(), Some("esterification"));
        assert_eq!(reaction.annotations["source_id"], "rxn-1");
    }

    #[test]
    fn full_graph_mode_is_explicitly_unsupported() {
        assert!(AskcosAdapter::full_graph().entries(output(), None).is_err());
    }

    #[test]
    fn direct_candidate_path_matches_entry_cast_across_worker_counts() {
        let adapter = AskcosAdapter::default();
        let entry = adapter.entries(output(), None).unwrap().pop().unwrap();
        let expected = adapter
            .cast(entry.payload, AdaptMode::Strict, None)
            .unwrap();

        for workers in [1, 2] {
            let candidates = adapt_candidates_with_workers(
                output(),
                &adapter,
                AdaptMode::Strict,
                None,
                None,
                None,
                workers,
            )
            .unwrap();
            assert_eq!(candidates.len(), 1);
            assert_eq!(
                serde_json::to_value(candidates[0].route.as_ref().unwrap()).unwrap(),
                serde_json::to_value(&expected).unwrap(),
            );
        }
    }
}
